File: src/main/java/io/sch/historyscan/infrastructure/features/startup/StartupDTO.java
package io.sch.historyscan.infrastructure.features.startup;

import com.fasterxml.jackson.annotation.JsonCreator;
import io.sch.historyscan.domain.error.HistoryScanTechnicalException;
import io.sch.historyscan.infrastructure.features.codebase.CodeBaseController;
import io.sch.historyscan.infrastructure.features.codebase.clone.CodeBaseToAddDTO;
import org.springframework.hateoas.RepresentationModel;
import org.springframework.http.HttpMethod;

import java.util.Objects;

import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;

public class StartupDTO extends RepresentationModel<StartupDTO> {

    private final String message;

    @JsonCreator
    public StartupDTO() throws HistoryScanTechnicalException {
        this.message = "Application is up! Please use the links below to navigate to the API endpoints.";
        try {
            addSelfLink();
            addCodebase();
            currentCodebases();
        } catch (NoSuchMethodException e) {
            throw new HistoryScanTechnicalException("No method found in the controller", e);
        }
    }

    private void currentCodebases() throws NoSuchMethodException {
        add(
                linkTo(CodeBaseController.class,
                        CodeBaseController.class.getMethod("currentCodeBases"))
                        .withRel("list-codebases")
                        .withTitle(HttpMethod.GET.name())
        );

    }

    private void addCodebase() throws NoSuchMethodException {
        add(
                linkTo(CodeBaseController.class,
                        CodeBaseController.class.getMethod("add", CodeBaseToAddDTO.class))
                        .withRel("add-codebase")
                        .withTitle(HttpMethod.POST.name())
        );

    }

    private void addSelfLink() throws NoSuchMethodException {
        add(
                linkTo(StartupController.class,
                        StartupController.class.getMethod("startup"))
                        .withSelfRel()
                        .withName("startup")
                        .withTitle(HttpMethod.GET.name())
        );
    }

    public String getMessage() {
        return message;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        StartupDTO that = (StartupDTO) o;
        return Objects.equals(message, that.message);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), message);
    }
}

File: historyscan-back/src/main/kotlin/io/sch/historyscan/infrastructure/features/startup/StartupDTO.kt
package io.sch.historyscan.infrastructure.features.startup

import com.fasterxml.jackson.annotation.JsonCreator
import io.sch.historyscan.domain.error.HistoryScanTechnicalException
import io.sch.historyscan.infrastructure.features.codebase.CodeBaseController
import io.sch.historyscan.infrastructure.features.codebase.clone.CodeBaseToAddDTO
import org.springframework.hateoas.RepresentationModel
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder
import org.springframework.http.HttpMethod
import java.util.*

class StartupDTO @JsonCreator constructor() : RepresentationModel<StartupDTO?>() {
    val message: String = "Application is up! Please use the links below to navigate to the API endpoints."

    init {
        try {
            addSelfLink()
            addCodebase()
            currentCodebases()
        } catch (e: NoSuchMethodException) {
            throw HistoryScanTechnicalException("No method found in the controller", e)
        }
    }

    @Throws(NoSuchMethodException::class)
    private fun currentCodebases() {
        add(WebMvcLinkBuilder.linkTo(CodeBaseController::class.java,
                CodeBaseController::class.java.getMethod("currentCodeBases"))
                .withRel("list-codebases")
                .withTitle(HttpMethod.GET.name())
        )
    }

    @Throws(NoSuchMethodException::class)
    private fun addCodebase() {
        add(WebMvcLinkBuilder.linkTo(CodeBaseController::class.java,
                CodeBaseController::class.java.getMethod("add", CodeBaseToAddDTO::class.java))
                .withRel("add-codebase")
                .withTitle(HttpMethod.POST.name())
        )
    }

    @Throws(NoSuchMethodException::class)
    private fun addSelfLink() {
        add(WebMvcLinkBuilder.linkTo(StartupController::class.java,
                StartupController::class.java.getMethod("startup"))
                .withSelfRel()
                .withName("startup")
                .withTitle(HttpMethod.GET.name())
        )
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        if (!super.equals(o)) return false
        val that = o as StartupDTO
        return message == that.message
    }

    override fun hashCode(): Int {
        return Objects.hash(super.hashCode(), message)
    }
}
